crates/cargo-util/src/process_builder.rs

```diff
@@ -340,9 +369,68 @@ impl ProcessBuilder {
         Ok(output)
     }
 
-    /// Converts `ProcessBuilder` into a `std::process::Command`, and handles the jobserver, if
-    /// present.
-    pub fn build_command(&self) -> Command {
+    /// Builds a command from `ProcessBuilder` and spawn it.
+    ///
+    /// There is a risk when spawning a process, it might hit "command line
+    /// too big" OS error. To handle those kind of OS errors, this method try
+    /// to reinvoke the command with a `@<path>` argfile that contains all the
+    /// arguments.
+    ///
+    /// * `apply`: Modify the command before invoking. Useful for updating [`Stdio`].
+    fn build_and_spawn(&self, apply: impl Fn(&mut Command)) -> io::Result<Child> {
+        let mut cmd = self.build_command();
+        apply(&mut cmd);
+
+        match cmd.spawn() {
+            Err(ref e) if self.should_retry_with_argfile(e) => {
+                let (mut cmd, _argfile) = self.build_command_with_argfile()?;
+                apply(&mut cmd);
+                cmd.spawn()
```

Comment:
Aren't we closing / removing the tempfile before the process has completed, racing with the application that is accessing it?

And should we explicitly close the tempfile, identifying any errors from the clean up?

Reply:
Fixed with 9155c0062bb6fcde2d8d727705841327dbe56e3f. It also removes the awkward `build_and_spawn` function as a side effect.